# 01 Importing libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [5]:
# Strings for path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [6]:
# Import Orders-Products-Prior data
df_ords_prior=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'))

In [7]:
# import cleaned Orders data 
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [8]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [9]:
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  first_order  
0                   2                  8                     NaN         True  
1                   3                  7                    15.0        False  
2                   3                 12                    21.0        False  
3                   4                  7                    29.0        False  
4                   4                 15                    28.0        False

In [10]:
df_ords.shape

(3421083, 9)

In [11]:
# Drop unnamed columns on Orders dataframe
df_ords=df_ords.drop(columns=['Unnamed: 0.1', 'Unnamed: 0'])

In [12]:
df_ords.shape

(3421083, 7)

In [13]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

In [14]:
df_ords_prior.shape

(32434489, 4)

In [15]:
# Check for data type on Orders-Products-Prior
df_ords_prior.dtypes

order_id             int64
product_id           int64
add_to_cart_order    int64
reordered            int64
dtype: object

In [16]:
df_ords_prior.describe()

order_id    product_id  add_to_cart_order     reordered
count  3.243449e+07  3.243449e+07       3.243449e+07  3.243449e+07
mean   1.710749e+06  2.557634e+04       8.351076e+00  5.896975e-01
std    9.873007e+05  1.409669e+04       7.126671e+00  4.918886e-01
min    2.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00
25%    8.559430e+05  1.353000e+04       3.000000e+00  0.000000e+00
50%    1.711048e+06  2.525600e+04       6.000000e+00  1.000000e+00
75%    2.565514e+06  3.793500e+04       1.100000e+01  1.000000e+00
max    3.421083e+06  4.968800e+04       1.450000e+02  1.000000e+00

In [18]:
# Optimize data type on on Orders-Products-Prior
df_ords_prior=df_ords_prior.astype({'order_id':'int32', 'product_id':'int32', 'add_to_cart_order':'int16', 'reordered':'int8'})

In [19]:
df_ords_prior.dtypes

order_id             int32
product_id           int32
add_to_cart_order    int16
reordered             int8
dtype: object

In [20]:
df_ords_prior.describe()

order_id    product_id  add_to_cart_order     reordered
count  3.243449e+07  3.243449e+07       3.243449e+07  3.243449e+07
mean   1.710749e+06  2.557634e+04       8.351076e+00  5.896975e-01
std    9.873007e+05  1.409669e+04       7.126671e+00  4.918886e-01
min    2.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00
25%    8.559430e+05  1.353000e+04       3.000000e+00  0.000000e+00
50%    1.711048e+06  2.525600e+04       6.000000e+00  1.000000e+00
75%    2.565514e+06  3.793500e+04       1.100000e+01  1.000000e+00
max    3.421083e+06  4.968800e+04       1.450000e+02  1.000000e+00

In [22]:
# Check for data type on Orders
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
first_order                  bool
dtype: object

In [23]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [25]:
# Optimize data type on Orders
df_ords=df_ords.astype({'order_id':'int32', 'user_id':'int32', 'order_number':'int16', 'orders_day_of_week':'int8', 'order_hour_of_day':'int8'})

In [26]:
df_ords.dtypes

order_id                    int32
user_id                     int32
order_number                int16
orders_day_of_week           int8
order_hour_of_day            int8
days_since_prior_order    float64
first_order                  bool
dtype: object

In [27]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Merging Dataframes 

In [29]:
# Merge Orders-Products-Prior and Orders dataframes
df_merged_large=df_ords.merge(df_ords_prior, on='order_id', indicator=True)

In [30]:
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [31]:
# Count frequency in _merge
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [32]:
# Merge Orders-Products-Prior and Orders dataframes using Outer join
df_merged_large_outer=df_ords.merge(df_ords_prior, on='order_id', how='outer', indicator=True)

In [33]:
df_merged_large_outer.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True       196.0                1.0   
1                     NaN         True     14084.0                2.0   
2                     NaN         True     12427.0                3.0   
3                     NaN         True     26088.0                4.0   
4                     NaN         True     26405.0                5.0   

   reordered _merge  
0        0.0   both  
1        0.0   both  
2        0.0   both  
3        0.0   both  
4        0.0   both

In [34]:
# Check frequency of _merge for outer joint
df_merged_large_outer['_merge'].value_counts()

_merge
both          32434489
left_only       206209
right_only           0
Name: count, dtype: int64

# After merging Orders-Products-Prior and Orders, the dataframe has 32,434,489 rows

# 04 Exporting data

In [37]:
df_merged_large.shape

(32434489, 11)

In [38]:
# Export merged Orders-Product dataframe
df_merged_large.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_product_combined.pkl')) 